File: URI/V1.2.6/python/src/rttp/pulse_header.py
"""
PulseHeader128 — RFC-002 §4.1 官方 128 字节硬件对齐帧头

（v1.3.0: 每个神经脉冲封装在临床级 PulseHeader128 中，
  128-BYTE 硬件对齐结构，与 CPU 双缓存行共振。）

字段布局（大端，与 RFC-002 §4.1 表逐一对应）:
    0x00  u32   RTTP_MAGIC    0x52545450（物理寄存器门验证）
    0x04  u128  VERSION_ID    锁定 130（v1.3.0-Alpha）
    0x14  u128  SEQUENCE_ID   单调脉冲序号（12ns 审计用）
    0x24  u128  TIMESTAMP     绝对纳秒发射时刻（12ns 精度）
    0x34  u8    TTL_PULSE     跳数上限（255 后脉冲蒸发）
    0x35  u8    PRIORITY      128-bit 分流权重（255 = Sovereign）
    0x36  u128  ROUTE_SHARD   12ns 抖动对齐的 Hive 导航哈希
    0x46  32B   AID_ORIGIN    256-bit 双分片身份 DNA（Genesis^Resonance）

--- v1.2.6 起：保留区分配（详见 SPEC/RTTP-FRAME-EXT-v1.2.6.md）---
    0x66  u8    SPEC_REV      0 = v1.2.6 之前（整块全零）; 1 = v1.2.6
    0x67  u8    FLAGS         bit0 = URI_ANCHORED; 其余位在 v1.2.6 MUST 为 0
    0x68  u8    ACTION_LEN    ACTION 有效字节数（0 = action 省略）
    0x69  16B   ACTION        §10.2 action 动词（小写 ASCII，右补 0x00）
    0x79  7B    RESERVED      v1.2.6 MUST 全零

⚠ **0x00–0x65 一个字节都没有改动**，VERSION_ID 亦保持 130。
   扩展只填「原本就是零」的保留区，因此只读 0x00–0x65 的旧实现天然兼容。
"""

import struct
import time

try:                        # pip 安装后：作为包内模块导入
    from . import rttp_uri
except ImportError:         # 平铺布局（DEMO/ 目录）：作为同目录脚本导入
    import rttp_uri         # action 语法（§10.2）唯一权威所在，避免规则重写

# ---- 固定区 ----
SIZE = 128
MAGIC = 0x52545450          # "RTTP"
VERSION_ID = 130            # v1.3.0-Alpha
AID_ORIGIN_OFFSET = 0x46

# ---- 扩展块（v1.2.6）----
SPEC_REV = 1

OFF_SPEC_REV = 0x66
OFF_FLAGS = 0x67
OFF_ACTION_LEN = 0x68
OFF_ACTION = 0x69
ACTION_MAX_LEN = 16
OFF_RESERVED = 0x79
RESERVED_LEN = 7

FLAG_URI_ANCHORED = 0x01

# 读端只认这些 SPEC_REV；未知修订一律拒斥（fail closed）
KNOWN_SPEC_REVS = (0, 1)


class PulseHeaderError(Exception):
    pass


# ---------------------------------------------------------------------------
# ACTION 编解码
# ---------------------------------------------------------------------------

def encode_action(action: str) -> bytes:
    """action 动词 → 16 字节（小写 ASCII，右补 0x00）。"""
    if action == "":
        return bytes(ACTION_MAX_LEN)
    if not rttp_uri.is_valid_action(action):
        raise PulseHeaderError(
            f"invalid action token {action!r} (RFC-002 §10.2: 1*( a-z / 0-9 / '-' ))")
    raw = action.encode("ascii")
    if len(raw) > ACTION_MAX_LEN:
        raise PulseHeaderError(
            f"action too long: {len(raw)} > {ACTION_MAX_LEN} bytes")
    return raw + bytes(ACTION_MAX_LEN - len(raw))


def decode_action(block16: bytes) -> str:
    """16 字节 ACTION 区 → 动词字符串（字节级解码，不含语法校验）。"""
    if len(block16) != ACTION_MAX_LEN:
        raise PulseHeaderError("action block must be 16 bytes")
    end = block16.find(b"\x00")
    raw = block16 if end < 0 else block16[:end]
    try:
        return raw.decode("ascii")
    except UnicodeDecodeError as exc:
        raise PulseHeaderError(f"ACTION is not ASCII: {exc}") from exc


# ---------------------------------------------------------------------------
# 构造
# ---------------------------------------------------------------------------

def build(sequence_id: int, ttl: int, priority: int,
          route_shard: bytes, aid_origin: bytes,
          timestamp_ns: int = None,
          action: str = "", uri_anchored: bool = False) -> bytes:
    """按官方解剖表构造 128 字节帧头。

    action / uri_anchored 为 v1.2.6 新增的**可选**参数：
      * 不传时产出与旧实现逐字节相同的帧（SPEC_REV 仍写 1，见 spec §4 R6）
      * action="" 表示 §10.4 的默认操作（standing read / 无动作谓词）
    """
    if len(route_shard) != 16:
        raise PulseHeaderError("route_shard must be 16 bytes (u128)")
    if len(aid_origin) != 32:
        raise PulseHeaderError("aid_origin must be 32 bytes (256-bit)")
    if not 0 <= sequence_id < (1 << 128):
        raise PulseHeaderError("sequence_id out of u128 range")
    if timestamp_ns is None:
        timestamp_ns = time.time_ns()

    action_block = encode_action(action)

    buf = bytearray(SIZE)
    struct.pack_into(">I", buf, 0x00, MAGIC)
    buf[0x04:0x14] = VERSION_ID.to_bytes(16, "big")
    buf[0x14:0x24] = sequence_id.to_bytes(16, "big")
    buf[0x24:0x34] = timestamp_ns.to_bytes(16, "big")
    buf[0x34] = ttl & 0xFF
    buf[0x35] = priority & 0xFF
    buf[0x36:0x46] = route_shard
    buf[0x46:0x66] = aid_origin

    # --- v1.2.6 扩展块 ---
    buf[OFF_SPEC_REV] = SPEC_REV
    buf[OFF_FLAGS] = FLAG_URI_ANCHORED if uri_anchored else 0x00
    buf[OFF_ACTION_LEN] = len(action.encode("ascii")) if action else 0
    buf[OFF_ACTION:OFF_ACTION + ACTION_MAX_LEN] = action_block
    # 0x79..0x80 保持零（RESERVED）

    return bytes(buf)


def build_for_uri(sequence_id: int, ttl: int, priority: int,
                  uri: str, aid_origin: bytes,
                  timestamp_ns: int = None) -> bytes:
    """从一条 `rttp` URI 直接构造帧头 —— 即 §10.4 所承诺的那次映射。

    ROUTE_SHARD 由 authority 派生（而非脉冲序号），ACTION 取自 path。
    """
    parsed = rttp_uri.parse(uri)
    return build(sequence_id, ttl, priority,
                 route_shard=parsed["route_shard"],
                 aid_origin=aid_origin,
                 timestamp_ns=timestamp_ns,
                 action=parsed["action"],
                 uri_anchored=True)


# ---------------------------------------------------------------------------
# 解析 / 校验
# ---------------------------------------------------------------------------

def parse(raw: bytes) -> dict:
    """解析并校验 128 字节帧头，返回字段字典（不判合法性，只做结构解码）。"""
    if not isinstance(raw, (bytes, bytearray)) or len(raw) != SIZE:
        raise PulseHeaderError(f"header must be exactly {SIZE} bytes")
    raw = bytes(raw)
    magic = struct.unpack_from(">I", raw, 0x00)[0]
    if magic != MAGIC:
        raise PulseHeaderError(f"bad RTTP_MAGIC: 0x{magic:08X}")

    spec_rev = raw[OFF_SPEC_REV]
    flags = raw[OFF_FLAGS]
    action_len = raw[OFF_ACTION_LEN]
    action = "" if spec_rev == 0 else decode_action(
        raw[OFF_ACTION:OFF_ACTION + ACTION_MAX_LEN])

    return {
        "magic_ok": True,
        "version_id": int.from_bytes(raw[0x04:0x14], "big"),
        "sequence_id": int.from_bytes(raw[0x14:0x24], "big"),
        "timestamp_ns": int.from_bytes(raw[0x24:0x34], "big"),
        "ttl": raw[0x34],
        "priority": raw[0x35],
        "route_shard": raw[0x36:0x46],
        "aid_origin": raw[0x46:0x66],
        # --- v1.2.6 ---
        "spec_rev": spec_rev,
        "flags": flags,
        "uri_anchored": bool(flags & FLAG_URI_ANCHORED),
        "action": action,
        "action_len": action_len,
        "action_omitted": action_len == 0,
    }


def verify(raw: bytes, expected_aid_origin: bytes = None) -> dict:
    """解析 + 校验版本、来源 AID 与扩展块，返回字段字典。

    扩展块校验规则（spec §4 R1–R7）：
      * SPEC_REV 不在 KNOWN_SPEC_REVS 内            → 拒斥
      * SPEC_REV=0 但扩展块非全零                    → 拒斥（形态自相矛盾）
      * SPEC_REV=1 但 RESERVED 非零 / FLAGS 未知位   → 拒斥
      * ACTION_LEN > 16 / 有内嵌 NUL / 非 §10.2 语法 → 拒斥
    """
    fields = parse(raw)
    if fields["version_id"] != VERSION_ID:
        raise PulseHeaderError(
            f"VERSION_ID mismatch: {fields['version_id']} != {VERSION_ID}")

    spec_rev = fields["spec_rev"]
    block = bytes(raw[OFF_SPEC_REV:SIZE])

    if spec_rev not in KNOWN_SPEC_REVS:
        raise PulseHeaderError(f"unknown SPEC_REV: {spec_rev} (fail closed)")
    if spec_rev == 0:
        if any(block):
            raise PulseHeaderError(
                "SPEC_REV=0 but extension block is not all-zero")
    else:
        if any(raw[OFF_RESERVED:SIZE]):
            raise PulseHeaderError("RESERVED bytes must be zero in SPEC_REV=1")
        unknown_flags = fields["flags"] & ~FLAG_URI_ANCHORED
        if unknown_flags:
            raise PulseHeaderError(
                f"unknown FLAGS bits set: 0x{unknown_flags:02X}")

        action_block = raw[OFF_ACTION:OFF_ACTION + ACTION_MAX_LEN]
        action = fields["action"]
        if fields["action_len"] > ACTION_MAX_LEN:
            raise PulseHeaderError(
                f"ACTION_LEN {fields['action_len']} > {ACTION_MAX_LEN}")
        if fields["action_len"] != len(action.encode("ascii")):
            raise PulseHeaderError("ACTION_LEN disagrees with ACTION payload")
        # 补零区必须全零（不留夹缝）
        if any(action_block[len(action.encode("ascii")):]):
            raise PulseHeaderError("ACTION padding must be zero")
        if action and not rttp_uri.is_valid_action(action):
            raise PulseHeaderError(f"ACTION violates RFC-002 §10.2: {action!r}")

    if expected_aid_origin is not None and \
            fields["aid_origin"] != expected_aid_origin:
        raise PulseHeaderError("AID_ORIGIN mismatch (spoofed origin?)")
    return fields


# ---------------------------------------------------------------------------
# 十六进制编解码
# ---------------------------------------------------------------------------

def to_hex(header_bytes: bytes) -> str:
    return header_bytes.hex()


def from_hex(hex_str: str) -> bytes:
    raw = bytes.fromhex(hex_str)
    if len(raw) != SIZE:
        raise PulseHeaderError("hex must decode to exactly 128 bytes")
    return raw

File: URI/V1.2.8/rust/src/pulse_header.rs
//! `PulseHeader128` -- RFC-002 sec. 4.1 official 128-byte hardware-aligned frame
//! header, plus the v1.2.6 extension block (`0x66`-`0x7F`).
//!
//! Field layout (big-endian, per sec. 4.1):
//! ```text
//!   0x00  u32   RTTP_MAGIC    0x52545450
//!   0x04  u128  VERSION_ID    locked at 130
//!   0x14  u128  SEQUENCE_ID   monotonic pulse sequence
//!   0x24  u128  TIMESTAMP     absolute nanosecond emission time
//!   0x34  u8    TTL_PULSE
//!   0x35  u8    PRIORITY
//!   0x36  u128  ROUTE_SHARD   SHA-256(authority)[0:16] -- see SPEC sec. 5
//!   0x46  32B   AID_ORIGIN    256-bit originator identity
//!   --- v1.2.6 extension (reserved area allocation, SPEC sec. 2) ---
//!   0x66  u8    SPEC_REV      0 = pre-v1.2.6 (whole block zero); 1 = v1.2.6
//!   0x67  u8    FLAGS         bit0 = URI_ANCHORED; others MUST be 0
//!   0x68  u8    ACTION_LEN    valid bytes of ACTION (0 = omitted)
//!   0x69  16B   ACTION        sec. 10.2 verb (lowercase ASCII, zero-padded)
//!   0x79  7B    RESERVED      MUST be zero
//! ```
//! Bytes `0x00`-`0x65` are UNTOUCHED and VERSION_ID stays 130, so readers of
//! the old layout keep working -- the extension only fills bytes that were
//! already zero.
//!
//! This is a real codec -- `build`/`verify` round-trip against the published
//! frame vectors -- in contrast to a raw memory reinterpretation.

use crate::rttp_uri::{self, RttpUriError};
use std::fmt;

pub const SIZE: usize = 128;
pub const MAGIC: u32 = 0x5254_5450; // "RTTP"
pub const VERSION_ID: u128 = 130;

pub const OFF_SPEC_REV: usize = 0x66;
pub const OFF_FLAGS: usize = 0x67;
pub const OFF_ACTION_LEN: usize = 0x68;
pub const OFF_ACTION: usize = 0x69;
pub const OFF_RESERVED: usize = 0x79;
pub const ACTION_MAX_LEN: usize = 16;
pub const RESERVED_LEN: usize = 7;

pub const FLAG_URI_ANCHORED: u8 = 0x01;
pub const SPEC_REV: u8 = 1;

/// Readers accept exactly these revisions; unknown ones fail closed.
pub const KNOWN_SPEC_REVS: [u8; 2] = [0, 1];

/// The only error type raised by the frame layer.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PulseHeaderError(pub String);

impl fmt::Display for PulseHeaderError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}
impl std::error::Error for PulseHeaderError {}

impl From<RttpUriError> for PulseHeaderError {
    fn from(e: RttpUriError) -> Self {
        PulseHeaderError(e.0)
    }
}

/// Decoded frame fields (structural read -- `verify` adds the semantics).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedFrame {
    pub version_id: u128,
    pub sequence_id: u128,
    pub timestamp_ns: u128,
    pub ttl: u8,
    pub priority: u8,
    pub route_shard: [u8; 16],
    pub aid_origin: [u8; 32],
    pub spec_rev: u8,
    pub flags: u8,
    pub uri_anchored: bool,
    pub action: String,
    pub action_len: u8,
    pub action_omitted: bool,
}

/// ACTION verb -> 16 bytes (lowercase ASCII, right-padded with 0x00).
pub fn encode_action(action: &str) -> Result<[u8; ACTION_MAX_LEN], PulseHeaderError> {
    let mut block = [0u8; ACTION_MAX_LEN];
    if action.is_empty() {
        return Ok(block);
    }
    if !rttp_uri::is_valid_action(action) {
        return Err(PulseHeaderError(format!(
            "invalid action token {:?} (RFC-002 sec. 10.2: 1*( a-z / 0-9 / '-' ))",
            action
        )));
    }
    let raw = action.as_bytes();
    if raw.len() > ACTION_MAX_LEN {
        return Err(PulseHeaderError(format!(
            "action too long: {} > {} bytes",
            raw.len(),
            ACTION_MAX_LEN
        )));
    }
    block[..raw.len()].copy_from_slice(raw);
    Ok(block)
}

/// 16-byte ACTION block -> verb string (byte-level decode, no syntax check).
pub fn decode_action(block16: &[u8]) -> Result<String, PulseHeaderError> {
    if block16.len() != ACTION_MAX_LEN {
        return Err(PulseHeaderError("action block must be 16 bytes".into()));
    }
    let end = block16.iter().position(|&b| b == 0).unwrap_or(ACTION_MAX_LEN);
    let raw = &block16[..end];
    match std::str::from_utf8(raw) {
        Ok(s) if s.is_ascii() => Ok(s.to_string()),
        _ => Err(PulseHeaderError("ACTION is not ASCII".into())),
    }
}

/// Build a 128-byte header per the official layout.
///
/// `action` / `uri_anchored` are the OPTIONAL v1.2.6 additions: omitting the
/// action produces a frame byte-identical to what an old implementation would
/// emit apart from SPEC_REV=1 (spec sec. 4 R6); empty action = sec. 10.4 default.
pub fn build(
    sequence_id: u128,
    ttl: u8,
    priority: u8,
    route_shard: &[u8; 16],
    aid_origin: &[u8; 32],
    timestamp_ns: u128,
    action: &str,
    uri_anchored: bool,
) -> Result<[u8; SIZE], PulseHeaderError> {
    let action_block = encode_action(action)?;

    let mut buf = [0u8; SIZE];
    buf[0x00..0x04].copy_from_slice(&MAGIC.to_be_bytes());
    buf[0x04..0x14].copy_from_slice(&VERSION_ID.to_be_bytes());
    buf[0x14..0x24].copy_from_slice(&sequence_id.to_be_bytes());
    buf[0x24..0x34].copy_from_slice(&timestamp_ns.to_be_bytes());
    buf[0x34] = ttl;
    buf[0x35] = priority;
    buf[0x36..0x46].copy_from_slice(route_shard);
    buf[0x46..0x66].copy_from_slice(aid_origin);

    // --- v1.2.6 extension block ---
    buf[OFF_SPEC_REV] = SPEC_REV;
    buf[OFF_FLAGS] = if uri_anchored { FLAG_URI_ANCHORED } else { 0x00 };
    buf[OFF_ACTION_LEN] = action.len() as u8;
    buf[OFF_ACTION..OFF_ACTION + ACTION_MAX_LEN].copy_from_slice(&action_block);
    // 0x79..0x80 stay zero (RESERVED)

    Ok(buf)
}

/// Build directly from an `rttp` URI -- the mapping sec. 10.4 promises:
/// ROUTE_SHARD derives from the authority, ACTION comes from the path.
pub fn build_for_uri(
    sequence_id: u128,
    ttl: u8,
    priority: u8,
    uri: &str,
    aid_origin: &[u8; 32],
    timestamp_ns: u128,
) -> Result<[u8; SIZE], PulseHeaderError> {
    let parsed = rttp_uri::parse(uri)?;
    build(
        sequence_id,
        ttl,
        priority,
        &parsed.route_shard,
        aid_origin,
        timestamp_ns,
        &parsed.action,
        true,
    )
}

/// Structural decode (no semantics).
pub fn parse(raw: &[u8]) -> Result<ParsedFrame, PulseHeaderError> {
    if raw.len() != SIZE {
        return Err(PulseHeaderError(format!("header must be exactly {} bytes", SIZE)));
    }
    let magic = u32::from_be_bytes([raw[0], raw[1], raw[2], raw[3]]);
    if magic != MAGIC {
        return Err(PulseHeaderError(format!("bad RTTP_MAGIC: 0x{:08X}", magic)));
    }
    let be_u128 = |r: &[u8]| -> u128 {
        let mut w = [0u8; 16];
        w.copy_from_slice(r);
        u128::from_be_bytes(w)
    };
    let spec_rev = raw[OFF_SPEC_REV];
    let action = if spec_rev == 0 {
        String::new()
    } else {
        decode_action(&raw[OFF_ACTION..OFF_ACTION + ACTION_MAX_LEN])?
    };
    let mut route_shard = [0u8; 16];
    route_shard.copy_from_slice(&raw[0x36..0x46]);
    let mut aid_origin = [0u8; 32];
    aid_origin.copy_from_slice(&raw[0x46..0x66]);
    let flags = raw[OFF_FLAGS];
    let action_len = raw[OFF_ACTION_LEN];

    Ok(ParsedFrame {
        version_id: be_u128(&raw[0x04..0x14]),
        sequence_id: be_u128(&raw[0x14..0x24]),
        timestamp_ns: be_u128(&raw[0x24..0x34]),
        ttl: raw[0x34],
        priority: raw[0x35],
        route_shard,
        aid_origin,
        spec_rev,
        flags,
        uri_anchored: flags & FLAG_URI_ANCHORED != 0,
        action_len,
        action_omitted: action_len == 0,
        action,
    })
}

/// Parse + validate version, origin AID and the extension block (rules R1-R7
/// of SPEC sec. 4). `expected_aid_origin` optionally pins the origin.
pub fn verify(
    raw: &[u8],
    expected_aid_origin: Option<&[u8; 32]>,
) -> Result<ParsedFrame, PulseHeaderError> {
    let fields = parse(raw)?;
    if fields.version_id != VERSION_ID {
        return Err(PulseHeaderError(format!(
            "VERSION_ID mismatch: {} != {}",
            fields.version_id, VERSION_ID
        )));
    }

    let spec_rev = fields.spec_rev;
    if !KNOWN_SPEC_REVS.contains(&spec_rev) {
        return Err(PulseHeaderError(format!(
            "unknown SPEC_REV: {} (fail closed)",
            spec_rev
        )));
    }

    if spec_rev == 0 {
        // SPEC_REV=0 demands the whole extension block be zero (formal
        // self-consistency -- R2).
        if raw[OFF_SPEC_REV..SIZE].iter().any(|&b| b != 0) {
            return Err(PulseHeaderError(
                "SPEC_REV=0 but extension block is not all-zero".into(),
            ));
        }
    } else {
        if raw[OFF_RESERVED..SIZE].iter().any(|&b| b != 0) {
            return Err(PulseHeaderError(
                "RESERVED bytes must be zero in SPEC_REV=1".into(),
            ));
        }
        let unknown = fields.flags & !FLAG_URI_ANCHORED;
        if unknown != 0 {
            return Err(PulseHeaderError(format!(
                "unknown FLAGS bits set: 0x{:02X}",
                unknown
            )));
        }

        let action_block = &raw[OFF_ACTION..OFF_ACTION + ACTION_MAX_LEN];
        let alen = fields.action_len as usize;
        if alen > ACTION_MAX_LEN {
            return Err(PulseHeaderError(format!(
                "ACTION_LEN {} > {}",
                alen, ACTION_MAX_LEN
            )));
        }
        if alen != fields.action.len() {
            return Err(PulseHeaderError(
                "ACTION_LEN disagrees with ACTION payload".into(),
            ));
        }
        // No gap between the verb and the padding (R5).
        if action_block[fields.action.len()..].iter().any(|&b| b != 0) {
            return Err(PulseHeaderError("ACTION padding must be zero".into()));
        }
        if !fields.action.is_empty() && !rttp_uri::is_valid_action(&fields.action) {
            return Err(PulseHeaderError(format!(
                "ACTION violates RFC-002 sec. 10.2: {:?}",
                fields.action
            )));
        }
    }

    if let Some(expected) = expected_aid_origin {
        if &fields.aid_origin != expected {
            return Err(PulseHeaderError("AID_ORIGIN mismatch (spoofed origin?)".into()));
        }
    }
    Ok(fields)
}

#[cfg(test)]
mod tests {
    use super::*;

    const AID: [u8; 32] = [0x42; 32];

    #[test]
    fn build_verify_roundtrip() {
        let raw = build(7, 255, 1, &[0xAA; 16], &AID, 1760000000, "verify", true).unwrap();
        let f = verify(&raw, Some(&AID)).unwrap();
        assert_eq!(f.version_id, 130);
        assert_eq!(f.spec_rev, 1);
        assert!(f.uri_anchored);
        assert_eq!(f.action, "verify");
        assert_eq!(f.route_shard, [0xAA; 16]);
    }

    #[test]
    fn spec_rev_zero_requires_all_zero_extension() {
        let mut raw = build(1, 64, 128, &[0u8; 16], &AID, 5, "", false).unwrap();
        raw[OFF_SPEC_REV] = 0;
        assert!(verify(&raw, None).is_ok());
        raw[OFF_ACTION] = 0x01; // contradicts SPEC_REV=0
        assert!(verify(&raw, None).is_err());
    }

    #[test]
    fn wrong_version_and_unknown_rev_fail() {
        let mut raw = build(1, 1, 1, &[0u8; 16], &AID, 1, "", false).unwrap();
        raw[0x04] = 0xFF;
        assert!(verify(&raw, None).is_err());
        let mut raw2 = build(1, 1, 1, &[0u8; 16], &AID, 1, "", false).unwrap();
        raw2[OFF_SPEC_REV] = 9;
        assert!(verify(&raw2, None).is_err());
    }

    #[test]
    fn action_rules() {
        assert!(encode_action("a-b-c9").is_ok());
        assert!(encode_action("-lead").is_err());
        assert!(encode_action("toolongverb_0123456789abcdef").is_err());
        let mut raw = build(1, 1, 1, &[0u8; 16], &AID, 1, "pulse", true).unwrap();
        raw[OFF_ACTION] = b'P'; // uppercase sneaks into ACTION
        assert!(verify(&raw, None).is_err());
    }
}
